library/cwm/src/lib/cwm/widget.rb

```diff
@@ -0,0 +1,450 @@
+    def fun_ref(*args)
+      Yast::FunRef.new(*args)
+    end
+
+  private
+
+    def init_method
+      fun_ref(method(:init), "void (string)")
+    end
+
+    def handle_method
+      fun_ref(method(:handle), "symbol (string, map)")
+    end
+
+    def store_method
+      fun_ref(method(:store), "void (string, map)")
+    end
+
+    def cleanup_method
+      fun_ref(method(:cleanup), "void (string)")
+    end
+
+    def validate_method
+      fun_ref(method(:validate), "boolean (string, map)")
+    end
+  end
+
+  # Represents custom widget, that have its UI content defined in method content.
+  # Useful mainly when specialized widget including more subwidget should be
+  # reusable at more places.
+  #
+  # @example custom widget child
+  #   class MyWidget < CWM::CustomWidget
+  #     def initialize
+  #       self.widget_id = "my_widget"
+  #     end
+  #
+  #     def content
+  #       HBox(
+  #         PushButton(Id(:reset), _("Reset")),
+  #         PushButton(Id(:undo), _("Undo"))
+  #       )
+  #     end
+  #
+  #     def handle(widget, event)
+  #       case event["ID"]
+  #       when :reset then ...
+  #       when :undo then ...
+  #       else ...
+  #       end
+  #     end
+  #   end
+  class CustomWidget < AbstractWidget
+    self.widget_type = :custom
+    # custom witget without content do not make sense
+    abstract_method :content
+
+    def description
+      {
+        "custom_widget" => content
+      }.merge(super)
+    end
+  end
+
+  # Empty widget useful mainly as place holder for replacement or for catching global events
+  #
+  # @example empty widget usage
+  #   widget = CWM::EmptyWidget("replace_point")
+  #   CWM.ShowAndRun(
+  #     "content" => VBox(widget.widget_id),
+  #     "widgets" => [widget]
+  #   )
+  class EmptyWidget < AbstractWidget
+    self.widget_type = :empty
+
+    def initialize(id)
+      self.widget_id = id
+    end
+  end
+
+  # helpers for easier set/obtain value of widget for widgets where value is
+  # obtained by :Value symbol
+  module ValueBasedWidget
+    def value
+      Yast::UI.QueryWidget(Id(widget_id), :Value)
+    end
+
+    def value=(val)
+      Yast::UI.ChangeWidget(Id(widget_id), :Value, val)
+    end
+  end
+
+  # helper to define items used by widgets that offer selection from list of
+  # values.
+  module ItemsSelection
+    # items are defined as list of pair, where first one is id and second
+    # one is user visible value
+    # @return [Array<Array<String>>]
+    # @example items method in widget
+    #   def items
+    #     [
+    #       [ "Canada", _("Canada")],
+    #       [ "USA", _("United States of America")],
+    #       [ "North Pole", _("Really cold place")],
+    #     ]
+    #   end
+    def items
+    end
+
+    def description
+      {
+        "items" => items
+      }.merge(super)
+    end
+
+    # change list of items offered in widget. Format is same as in {#items}
+    def change_items(items_list)
```

Comment:
Aaah, now I see. So that would be `items`, `items=` doing a QueryWidget/ChangeWidget and a `default_items` setting the `description`... which I still don't like much better.